File: LeetCode - Algorithm/Hash/128.js
// 128. Longest Consecutive Sequence
// https://leetcode.com/problems/longest-consecutive-sequence/

/**
 * @param {number[]} nums
 * @return {number}
 */
var longestConsecutive = function (nums) {
  let result = 0;

  const numsSet = new Set(nums);

  for (let i = 0; i < nums.length; i++) {
    if (numsSet.has(nums[i] - 1)) continue;

    let currentNum = nums[i];
    let currentCount = 0;
    while (numsSet.has(currentNum)) {
      currentCount += 1;
      currentNum += 1;
    }

    result = Math.max(result, currentCount);
  }

  return result;
};

`
nums = [100,4,200,1,3,2]

1, 2, 3, 4, 100, 200
`;

File: LeetCode - Algorithm/Hash/128.java
// 128. Longest Consecutive Sequence
// https://leetcode.com/problems/longest-consecutive-sequence/

import java.util.Set;
import java.util.HashSet;
import java.util.List;
import java.util.LinkedList;

class Solution {
    // Time Limit Exceeded :(
    // time complexity: O(n^2)
    public int longestConsecutive1(int[] nums) {
        Set<Integer> numsSet = new HashSet<>();
        for (int num : nums)
            numsSet.add(num);

        int result = 0;

        for (int num : nums) {
            int currentCount = 1;

            int target = num + 1;
            while (numsSet.contains(target)) {
                currentCount += 1;

                target += 1;
            }

            target = num - 1;
            while (numsSet.contains(target)) {
                currentCount += 1;

                target -= 1;
            }

            result = Math.max(result, currentCount);
        }

        return result;
    }

    // time complexity: O(3n) = O(n)
    public int longestConsecutive2(int[] nums) {
        Set<Integer> numsSet = new HashSet<>();
        for (int num : nums)
            numsSet.add(num);

        List<Integer> maxValues = new LinkedList<>();

        for (int num : nums) {
            if (numsSet.contains(num - 1))
                continue;

            maxValues.add(num);
        }

        int result = 0;

        for (int i = 0; i < maxValues.size(); i++) {
            int currentCount = 0;

            int target = maxValues.get(i);
            while (numsSet.contains(target)) {
                currentCount += 1;

                target += 1;
            }

            result = Math.max(result, currentCount);
        }

        return result;
    }
}

/*
 * Even though the second method has a time complexity of O(n), the LeetCode
 * system is returning a "Time Limit Exceeded" message. :( I understand getting
 * a "Time Limit Exceeded" message for the first method since its time
 * complexity is O(n^2), but the second method should be accepted. :(
 */